### PR TITLE
docs(s3): updated docstring for BucketPolicy()

### LINKED ISSUE #31358: aws-s3: The `autoDeleteObjects` prop of `Bucket` is too brittle
### Describe the bug

We are going to deploy this CDK stack and observe the behavior.

```ts
const uniqueId = this.node.addr.substring(0, 8);

const myBucket = new Bucket(this, "MyBucket", {
  bucketName: `my-bucket-${uniqueId}`,
  enforceSSL: true,
  removalPolicy: RemovalPolicy.DESTROY,
  autoDeleteObjects: true,
});

const myBucketPolicy = new BucketPolicy(this, "MyBucketPolicy", {
  bucket: myBucket,
});

const myUser = new User(this, "MyUser", {
  userName: "my-user",
});

myBucketPolicy.document.addStatements(
  new PolicyStatement({
    effect: Effect.ALLOW,
    principals: [myUser],
    actions: ["s3:GetObject", "s3:ListBucket"],
    resources: [myBucket.bucketArn, `${myBucket.bucketArn}/*`],
  }),
);
```

### Regression Issue

- [ ] Select this option if this issue appears to be a regression.

### Last Known Working CDK Version

_No response_

### Expected Behavior

I expected the `cdk deploy` to fail immediately due to a conflict between the `autoDeleteObjects` custom resource and the `BucketPolicy` we defined. Since only one bucket policy can exist at a time, the system should have detected this conflict and prevented the deployment.

### Current Behavior

However, the stack deploys successfully without any error.

image

Inspecting the S3 bucket policy shows that the `BucketPolicy` we explicitly created overrides the one generated by `autoDeleteObjects`:

```json
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Principal": {
                "AWS": "arn:aws:iam::533267016779:user/my-user"
            },
            "Action": [
                "s3:GetObject",
                "s3:ListBucket"
            ],
            "Resource": [
                "arn:aws:s3:::my-bucket-c812be60",
                "arn:aws:s3:::my-bucket-c812be60/*"
            ]
        }
    ]
}
```

When attempting to destroy the stack with `cdk destroy`, it fails because the Lambda function linked to `autoDeleteObjects` custom resource lacks the necessary permissions to delete the S3 bucket. This occurs because the `BucketPolicy` we defined overrides the policy that grants the required permissions.

### Reproduction Steps

see above

### Possible Solution

The CDK should enforce a check to prevent defining a `BucketPolicy` that overrides the permissions needed by the `autoDeleteObjects` custom resource. Although the deployment works, it leads to conflicts during stack destruction due to missing permissions.

### Additional Information/Context

_No response_

### CDK CLI Version

2.156.0

### Framework Version

_No response_

### Node.js Version

20.17.0

### OS

Ubuntu 22.04.3 LTS

### Language

TypeScript

### Language Version

_No response_

### Other information

_No response_

### PR DESCRIPTION
**Pull request by khushail**
### Issue #[31358](https://github.com/aws/aws-cdk/issues/31358)

Closes #31358 .

### Reason for this change

Exsiting [CDK Doc on BucketPolicy](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_s3.BucketPolicy.html) mentions as `Prefer to use` which is misleading as it does not clearly states the reprecussions. 

### Description of changes

I have added a sample of what would happen if this is used along with other Bucket properties.

### Description of how you validated changes

This is a minor documentation change

### Checklist
- [ ] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by khushail** [comment]
Exemption Request

File: packages/aws-cdk-lib/aws-s3/lib/bucket-policy.ts [hunk 1]
```diff
@@ -25,13 +25,26 @@ export interface BucketPolicyProps {
  * Policies define the operations that are allowed on this resource.
  *
  * You almost never need to define this construct directly.
- *
+ *
```
Comment (kaizencc) [inline packages/aws-cdk-lib/aws-s3/lib/bucket-policy.ts:28]:
```suggestion
 *
```

File: packages/aws-cdk-lib/aws-s3/lib/bucket-policy.ts [hunk 2]
```diff
@@ -25,13 +25,26 @@ export interface BucketPolicyProps {
...
+ * added to the bucket policy: 
+ *    ["s3:DeleteObject*", "s3:GetBucket*", "s3:List*", "s3:PutBucketPolicy"], 
+ * and when you add a new BucketPolicy with ["s3:GetObject", "s3:ListBucket"] on 
+ * this existing bucket, invoking `BucketPolicy()` will create a new Policy 
+ * without knowing one earlier exists already, so it creates a new one. 
+ * In this case, the custom resource handler will not have access to 
+ * `s3:GetBucketTagging` action which will cause failure during deletion of stack.
+ * 
+ * Hence its strongly recommended to use `addToResourcePolicy()` method to add 
+ * new permissions to existing policy.
```
Comment (kaizencc) [inline packages/aws-cdk-lib/aws-s3/lib/bucket-policy.ts:46]:
```suggestion
 * The bucket policy method is implemented differently than `addToResourcePolicy()`
 * as `BucketPolicy()` creates a new policy without knowing one earlier existed.
 * e.g. if during Bucket creation, if `autoDeleteObject:true`, these policies are
 * added to the bucket policy:
 *    ["s3:DeleteObject*", "s3:GetBucket*", "s3:List*", "s3:PutBucketPolicy"],
 * and when you add a new BucketPolicy with ["s3:GetObject", "s3:ListBucket"] on
 * this existing bucket, invoking `BucketPolicy()` will create a new Policy
 * without knowing one earlier exists already, so it creates a new one.
 * In this case, the custom resource handler will not have access to
 * `s3:GetBucketTagging` action which will cause failure during deletion of stack.
 *
 * Hence its strongly recommended to use `addToResourcePolicy()` method to add
 * new permissions to existing policy.
```

**Review by kaizencc — DISMISSED** [review]

File: packages/aws-cdk-lib/aws-s3/lib/bucket-policy.ts [hunk 3]
```diff
@@ -30,8 +30,21 @@ export interface BucketPolicyProps {
...
+ * added to the bucket policy:
+ *    ["s3:DeleteObject*", "s3:GetBucket*", "s3:List*", "s3:PutBucketPolicy"],
+ * and when you add a new BucketPolicy with ["s3:GetObject", "s3:ListBucket"] on
+ * this existing bucket, invoking `BucketPolicy()` will create a new Policy
+ * without knowing one earlier exists already, so it creates a new one.
+ * In this case, the custom resource handler will not have access to
+ * `s3:GetBucketTagging` action which will cause failure during deletion of stack.
+ *
+ * Hence its strongly recommended to use `addToResourcePolicy()` method to add
+ * new permissions to existing policy.
```
Comment (kaizencc) [inline packages/aws-cdk-lib/aws-s3/lib/bucket-policy.ts:46]:
```suggestion
 *
 * The bucket policy method is implemented differently than `addToResourcePolicy()`
 * as `BucketPolicy()` creates a new policy without knowing one earlier existed.
 * e.g. if during Bucket creation, if `autoDeleteObject:true`, these policies are
 * added to the bucket policy:
 *    ["s3:DeleteObject*", "s3:GetBucket*", "s3:List*", "s3:PutBucketPolicy"],
 * and when you add a new BucketPolicy with ["s3:GetObject", "s3:ListBucket"] on
 * this existing bucket, invoking `BucketPolicy()` will create a new Policy
 * without knowing one earlier exists already, so it creates a new one.
 * In this case, the custom resource handler will not have access to
 * `s3:GetBucketTagging` action which will cause failure during deletion of stack.
 *
 * Hence its strongly recommended to use `addToResourcePolicy()` method to add
 * new permissions to existing policy.
```

**Review by kaizencc — APPROVED** [review]
you keep missing linter errors on this doc change :). you can run `cdk lint --fix` on your local setup to make sure that it passes the linter before the PR stage.

**Comment by khushail** [comment]
> you keep missing linter errors on this doc change :). you can run `cdk lint --fix` on your local setup to make sure that it passes the linter before the PR stage.

sounds good. will keep in mind for the next time. :)

**Pull request merged**